### PR TITLE
Issue #277 - destDir optional

### PR DESCRIPTION
**Pull request by cverbiest**
Make destDir optional parameter
Added tests/PCTCompile/test65/build.xml but don't know how to turn into real test

### CODE REVIEW

File: src/java/com/phenix/pct/PCTCompile.java [hunk 1]
```diff
@@ -277,13 +279,11 @@ private boolean createDir(File dir) {
...
         // Create dest directory if necessary
-        if (compAttrs.getDestDir() == null) {
-            cleanup();
-            throw new BuildException(Messages.getString("PCTCompile.34")); //$NON-NLS-1$
-        }
-        if (!createDir(compAttrs.getDestDir())) {
-            cleanup();
-            throw new BuildException(MessageFormat.format(Messages.getString("PCTCompile.36"), "destDir")); //$NON-NLS-1$
+        if (compAttrs.getDestDir() != null) {
+            if (!createDir(compAttrs.getDestDir())) {
```
Comment (jakejustus) [inline src/java/com/phenix/pct/PCTCompile.java]: MAJOR Merge this if statement with the enclosing one. [rule](http://sonar.riverside-software.fr/coding_rules#rule_key=squid%3AS1066)

**Comment by jakejustus** [comment]
SonarQube analysis reported 2 issues
* ![BLOCKER][BLOCKER] 1 blocker
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![BLOCKER][BLOCKER] [compile.p#L371](https://github.com/Riverside-Software/pct/blob/76ded6d224aa74982f105ad26fbf29d710a8418e/src/progress/pct/compile.p#L371): unexpected token: OPTIONS in src/progress/pct/compile.p:371:90 [rule](http://sonar.riverside-software.fr/coding_rules#rule_key=rssw-oe%3Aproparse.error)

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by jakejustus** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![BLOCKER][BLOCKER] [compile.p#L371](https://github.com/Riverside-Software/pct/blob/01d976bfd6b4735110fb6c461d48e5f70781284e/src/progress/pct/compile.p#L371): unexpected token: OPTIONS in src/progress/pct/compile.p:371:90 [rule](http://sonar.riverside-software.fr/coding_rules#rule_key=rssw-oe%3Aproparse.error)

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'

**Comment by gquerret** [comment]
Did you change code to remove the `INTO` option ?

**Comment by cverbiest** [comment]
Yes, I found that save into ? (Unknown) has the same effect as not specifying into.
It's was easier to change it that way.

It's working for me but I'm still working on the tests.
The purpose of Test1 seems to be to check if pctcompile fails if no destDir is specified. Should I remove that test altogether or try to fix it ?

**Comment by gquerret** [comment]
Test1 won't be used anymore if destDir is not mandatory, just remove it.

**Comment by jakejustus** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![BLOCKER][BLOCKER] [compile.p#L371](https://github.com/Riverside-Software/pct/blob/5671fc519259b86e3341c7f79863f254bf5d25a6/src/progress/pct/compile.p#L371): unexpected token: OPTIONS in src/progress/pct/compile.p:371:90 [rule](http://sonar.riverside-software.fr/coding_rules#rule_key=rssw-oe%3Aproparse.error)

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'

File: src/progress/pct/compile.p [hunk 2]
```diff
@@ -81,6 +81,7 @@ DEFINE STREAM sWarnings.
 
 /* PCTCompile attributes */
 DEFINE VARIABLE OutputDir AS CHARACTER  NO-UNDO.
+DEFINE VARIABLE DestDir AS CHARACTER  NO-UNDO INITIAL ?.
```
Comment (gquerret) [inline src/progress/pct/compile.p:84]: Just wondering, why don't you re-use outputDir variable ?

Comment (cverbiest) [reply src/progress/pct/compile.p:84]: I started out using OutputDir but a number of variables are derived from OutputDir.
The change was a lot smaller and IMHO less likely to introduce errors if I left OutputDir as a valid directory.

Revisiting the code I see that I probably missed 

    /* Does .r file exists ? */
    ASSIGN RCodeTS = getTimeStampDF(OutputDir, ipOutFile).

Comment (gquerret) [reply src/progress/pct/compile.p:84]: > The change was a lot smaller and IMHO less likely to introduce errors if I left OutputDir as a valid directory.

Good idea

> Revisiting the code I see that I probably missed...

Missed what ?

**Comment by gquerret** [comment]
I'll make a few changes later this week before integrating the PR, mostly in order to use test case from branch Issue139 (this specific issue was raised some time ago, and I created this test case hoping the reporter would be able to provide a reproducible test case, but that didn't happen). I'll also add another test case to see if that's not a problem when compiling procedural code (or if it should be mentioned in the documentation that problems may occur when destDir is not specified).

**Comment by cverbiest** [comment]
Ok, I'll do some more changes : 
* rename OUTPUTDIR to DESTDIR in PCTCompile.java and compile.p to avoid confusion.
* I'm pretty sure that getting the rcode timestamp fails and that sources are always compiled, I'll fix that

Is there a test case that fails if unchanged files are recompiled ?

File: src/progress/pct/compile.p [hunk 3]
```diff
@@ -381,6 +383,7 @@ PROCEDURE compileXref.
   END.
 &ENDIF
 
+  IF DestDir EQ ? THEN cSaveDir = ?.
```
Comment (gquerret) [inline src/progress/pct/compile.p]: Would move this statement to line 284 where cSaveDir is assigned

**Comment by gquerret** [comment]
> Is there a test case that fails if unchanged files are recompiled ?

Yes, 100% sure on procedural code, but that's probably also tested on OO code

**Comment by cverbiest** [comment]
I reverted the OUTPUTDIR change it caused a lot of failed tests. It wasn't clear to me why it caused those failures.
RCodeTS is now corrected

**Comment by gquerret** [comment]
Unit tests on my Jenkins server are successful for commit 06a210b

File: src/progress/pct/compile.p [hunk 4]
```diff
@@ -300,8 +302,8 @@ PROCEDURE compileXref.
     ASSIGN opComp = 5.
   END.
   ELSE DO:
-    /* Does .r file exists ? */
-    ASSIGN RCodeTS = getTimeStampDF(OutputDir, ipOutFile).
+    /* Does .r file exists ?, if DestDir = unknown rcode will be located in the same directory as the source : ipInDir */
```
Comment (jakejustus) [inline src/progress/pct/compile.p]: MINOR Line is 121 characters long, exceeding maximum value of 120 [rule](http://sonar.riverside-software.fr/coding_rules#rule_key=rssw-oe-main%3Aeu.rssw.antlr.proparse.checks.LinesTooLong)

File: src/progress/pct/compile.p [hunk 5]
```diff
@@ -281,7 +283,7 @@ PROCEDURE compileXref.
   IF (opError) THEN RETURN.
   ASSIGN opError = NOT createDir(PCTDir, cBase).
   IF (opError) THEN RETURN.
-  cSaveDir = IF cFileExt = ".cls" OR lRelative THEN outputDir ELSE outputDir + '/':U + cBase.
+  cSaveDir = IF DestDir EQ ? THEN ? ELSE (IF cFileExt = ".cls" OR lRelative THEN outputDir ELSE outputDir + '/':U + cBase).
```
Comment (jakejustus) [inline src/progress/pct/compile.p]: MINOR Line is 123 characters long, exceeding maximum value of 120 [rule](http://sonar.riverside-software.fr/coding_rules#rule_key=rssw-oe-main%3Aeu.rssw.antlr.proparse.checks.LinesTooLong)

**Comment by jakejustus** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by gquerret** [comment]
I've merged a breaking change in pctCompile this week-end. I can do the rebase if you want, that might be faster for me.

**Comment by cverbiest** [comment]
@gquerret I'm ready with that branch you can rebase

**Comment by gquerret** [comment]
Rebased (and squashed) your changes, then added test case for PCTCompileExt + changes to PCTCompileExt + another test case coming from issue #139 
Hopefully that will be enough.

**Pull request merged**